### PR TITLE
Migrate hawkular metrics to a new namespace

### PR DESCRIPTION
**Pull request by rubenvp8510**

### CODE REVIEW

File: roles/openshift_metrics/tasks/install_metrics.yaml [hunk 1]
```diff
@@ -1,6 +1,41 @@
...
+    namespace: "{{ openshift_metrics_old_project }}"
+  register: openshift_metrics_infra
+
+- name:  Generate  template for project {{ openshift_metrics_project }} creation
+  template:
+    src: create_project.j2
+    dest: "{{ mktemp.stdout }}/templates/hawkular-metrics-projects.yaml"
+  vars:
+    project_name: "{{ openshift_metrics_project }}"
+    annotations: "{{ openshift_metrics_infra.results.results[0].metadata.annotations }}"
```
Comment (jsanda) [inline roles/openshift_metrics/tasks/install_metrics.yaml]: One of the things I have been wondering is whether or not we should copy all of the annotations or just the 3 that we need. Here are the annotations from an lonline cluster:

```yaml
kind: Project
metadata:
  annotations:
    openshift.io/description: openshift-infra
    openshift.io/display-name: openshift-infra
    openshift.io/logging.ui.hostname: kibana-ops.c1b9.mycluster.openshiftapps.com
    openshift.io/node-selector: type=infra
    openshift.io/sa.initialized-roles: "true"
    openshift.io/sa.scc.mcs: s0:c3,c2
    openshift.io/sa.scc.supplemental-groups: 1000010000/10000
    openshift.io/sa.scc.uid-range: 1000010000/10000
```
I am not sure what the hostname annotation is before but I suspect that is something that does not get by the installer. I also wonder about the description and display-name annotations. If we do want those, I imagine they should refer to the new namespace. We can ask someone from the logging team about this.

Comment (rubenvp8510) [reply roles/openshift_metrics/tasks/install_metrics.yaml]: That is true, we only need 3 annotations according to this comment https://github.com/openshift/openshift-ansible/pull/8649#discussion_r193788678

If this PR works for you I can update the PR for 3.10 and we can ask.

I do not think is a big deal to filter this.

Comment (rubenvp8510) [reply roles/openshift_metrics/tasks/install_metrics.yaml]: Indeed I would prefer to set only those three. I feel more safe :)

**Comment by jsanda** [comment]
Unfortunately, I am not familiar enough with oc_apply.yaml to offer any good insights at the moment. @rubenvp8510 can you summarize the changes in it? In particular, I am curious about the blocks that have:

```yaml
when: namespace is not defined
```

File: roles/openshift_metrics/tasks/migrate_cassandra_pvcs.yaml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        patch pv {{ openshift_metrics_pv_name }} -p '{"spec": {"claimRef": { "namespace": "{{ openshift_metrics_project }}", "name": "{{ openshift_metrics_cassandra_pvc_prefix }}-{{ metrics_pvc_index }}", "uid": "{{ openshift_metrics_pvc_uuid }}" }}}'
+
+    - name: Wait until the PV is attached to new PVC
+      oc_obj:
+        state: list
+        kind: pvc
+        name: "{{ openshift_metrics_cassandra_pvc_prefix }}-{{ metrics_pvc_index }}"
+        namespace: "{{openshift_metrics_project}}"
+      register: openshift_metrics_new_pvc
+      until: openshift_metrics_new_pvc.results.results[0].spec.volumeName == openshift_metrics_pv_name
```
Comment (jsanda) [inline roles/openshift_metrics/tasks/migrate_cassandra_pvcs.yaml:102]: I think we should update the ```until``` condition to also check that the status.phase of the new PVC is Bound.

File: roles/openshift_metrics/tasks/generate_cassandra_pvcs.yaml [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
+  include_tasks: migrate_cassandra_pvcs.yaml
 
 # _metrics_pvc.results.results | length > 0 returns a false positive
 # so we check for the presence of 'stderr' to determine if the obj exists or not
 # the RC for existing and not existing is both 0
+
+# No PVC exist (either new or old namespace), so it's a new installation
 - when:
-    - _metrics_pvc.results.stderr is defined
+    - metrics_pvc.results.stderr is defined and not openshift_metrics_migrate_namespace
```
Comment (jsanda) [inline roles/openshift_metrics/tasks/generate_cassandra_pvcs.yaml:20]: I think this check is good. We proceed in this block only if neither the PVC exists in the new namespace nor in the old namespace.

**Comment by jsanda** [comment]
I am still testing the changes, but I think you should go ahead and update the upstream PR with the changes. The sooner the better. Then I will copy over my new comments. Nice work!

**Comment by jsanda** [comment]
@rubenvp8510 I just finished my smoke testing which included installing and uninstalling from both namespaces. Everything looks good so far :)
